src/community/label_propagation.jl

```diff
@@ -0,0 +1,99 @@
+"""
+Community detection using the label propagation algorithm (see [Raghavan et al.](http://arxiv.org/abs/0709.2938)).
+`g`: imput Graph
+`maxiter`: optional, maximum iterations
+return : array containing vertex assignments, the range of the output as 1:countdistinct(membership)
+"""
+function label_propagation(g::SimpleGraph; maxiter=1000)
+    n = nv(g)
+    membership = collect(1:n)
+    active_nodes = Set(vertices(g))
+    nc = NeighComm(collect(1:n), fill(-1,n), 1)
+    i = 0
+    while !isempty(active_nodes) && i < maxiter
+    	i += 1
+        random_order = shuffle(collect(active_nodes))
+        for u in random_order
+          old_comm = membership[u]
+          membership[u] = vote!(g, membership, nc, u)
+            if old_comm != membership[u]
+                for v in out_neighbors(g, u)
+                    push!(active_nodes, v)
+                end
+            else
+                delete!(active_nodes, u)
+            end
+        end
+    end
+    renumber_labels!(membership)
+    membership
+end
+
+"""Fast shuffle Array `a` in UnitRange `r` inplace."""
+function range_shuffle!(r::UnitRange, a::AbstractVector)
+	for i=length(r):-1:2
+		j = rand(1:i)
+		ii = i + r.start - 1
+		jj = j + r.start - 1
+		a[ii],a[jj] = a[jj],a[ii]
+	end
+end
+
+"""Type to record neighbor labels and their counts."""
+type NeighComm
+	neigh_pos::Vector{Int}
+	neigh_cnt::Vector{Int}
+	neigh_last::Int
+end
+
+"""Return the most frequency label."""
+function vote!(g::SimpleGraph, membership::Vector{Int}, nc::NeighComm, u::Int)
+    for i=1:nc.neigh_last-1
+        nc.neigh_cnt[nc.neigh_pos[i]] = -1
+    end
+    nc.neigh_last = 1
+    nc.neigh_pos[1] = membership[u]
+    nc.neigh_cnt[nc.neigh_pos[1]] = 0
+    nc.neigh_last = 2
+    max_cnt = 0
+    for neigh in out_neighbors(g,u)
+        neigh_comm = membership[neigh]
+        if nc.neigh_cnt[neigh_comm] < 0
+            nc.neigh_cnt[neigh_comm] = 0
+            nc.neigh_pos[nc.neigh_last] = neigh_comm
+            nc.neigh_last += 1
+        end
+        nc.neigh_cnt[neigh_comm] += 1
+        if nc.neigh_cnt[neigh_comm] > max_cnt
+          max_cnt = nc.neigh_cnt[neigh_comm]
+        end
+    end
```

Comment:
So this structure is used as a scratch space for storing per neighbor information?

So the way to read this is that 
```
label[neighbors(g,v)[1]] == 1
label[neighbors(g,v)[2]] == 2
label[neighbors(g,v)[3]] == 4
```

Is there a typo in your example of `neigh_cnt`? If not what are the `-1, -1, 2`? I would expect `[1,1,-1,1]` given the rest of the example.

And neigh_last is holding the length of the neighborhood not maximum(neighbors(g,v))?